### ISSUE TITLE
Concerns about `Authenticator._validate_authenticator()`

### ISSUE THREAD

**Issue by nmtake** [praw-dev/prawcore#44]
Commit 342e3e8 introducees `Authenticator._validate_authenticator()`. If I understand correctly, that validator prevents users to use `code` grant flow from `installed` clients. Since many distributed desktop applications uses `code` grant flow (with optional refresh token), I believe that validator could be a problem.

**Comment by bboe** [comment]
It would seem my understanding of the installed client type is incomplete. What do you propose we do to fix and support all the use cases at this point?

**Comment by bboe** [comment]
I took a look through the documents again and I should have a better understanding now.

The "installed" app as far as PRAW is concerned is missing the part at the moment to obtain a token on behalf of another user. In fact, the authorize_url that is part of the Authenticator is actually wrong for the installed app type as its requested `response_type` should be `token`, not `code`.

For this purpose, I think it would make sense to split into separate Authenticators. `TrustedAuthenticator` that has a client secret, and `UntrustedAuthentictor`, which does not not take a client secret (TrustedAuthenticator could inherit from UntrustedAuthentictor).

The DeviceIDAuthorizer, then can validate that only UntrustedAuthenticator is being used. However, DeviceIDAuthorizer in simpler terms appears to be a read-only authorizer when using the installed app. We probably don't need to change things in prawcore, but I think I'll use the read-only concept to interact with this Authorizer in PRAW.

Finally, what is missing is the ability to manually input a received token for the implicit grant flow. For this it would seem to make sense to add an ImplicitAuthorizer that directly takes a token, an UntrustedAuthorizer, and does not support refresh. I'll see what I can come up with for these. Thanks for bringing it to my attention.

**Comment by bboe** [comment]
Concerns should be addressed by #45. Please reopen, or create a new issue if the concerns were not addressed, or there are new concerns. Thanks again!

**Comment by nmtake** [comment]
I took a look at #45, but it seems that the commit didn't address above concerns. Following code raises `prawcore.exceptions.InvalidInvocation: Must use a authenticator of type TrustedAuthenticator.` prawcore version is 0.2.1.

```python
import os
import prawcore

authenticator = prawcore.UntrustedAuthenticator(
    prawcore.Requestor('code_grant_test_from_installed_client'),
    os.environ['PRAWCORE_CLIENT_ID'])
authorizer = prawcore.Authorizer(authenticator)
```

So installed clients cannot use code grant flow in prawcore v0.2.1.  

IMO setting `UntrustedAuthenticator.RESPONSE_TYPE` to `'token'` is wrong since installed clients should be able to use not only [implicit auth flow](https://github.com/reddit/reddit/wiki/OAuth2#authorization-implicit-grant-flow) but also [code auth flow](https://github.com/reddit/reddit/wiki/OAuth2#authorization).

PS. To demonstrate installed clients can obtain acceess token (and refresh token) with code grant flow, [I wrote this gist. ](https://gist.github.com/nmtake/251b6ece0965667334a671ab13b5be7b)
